job-runner/chain/reaper.go

```diff
@@ -0,0 +1,630 @@
+			r.runJobChan <- sequenceStartJob // re-enqueue first job in sequence
+			return
+		}
+
+		// If the failed job was stopped or is not part of a retryable
+		// sequence, then ignore subsequent jobs. For example, if job B of
+		// A -> B -> C  fails, then C is not run.
+		jLogger.Warn("job did not complete successfully.")
+		return
+	}
+
+	jLogger.Infof("job completed successfully.")
+	for _, nextJob := range r.chain.NextJobs(job.Id) {
+		nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJob.Id})
+
+		// Copy job data to every child job, even if it's not ready to be run yet.
+		// When a job has multiple parent jobs, it'll get job data copied from each
+		// parent, not just the last one to finish. Be careful - it's possible for
+		// parents to overwrite each other's job data if they set the same field.
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
+			nextJLogger.Infof("next job is not runnable - not enqueuing it")
+			continue
+		}
+		nextJLogger.Infof("next job is runnable - enqueuing it")
+		if r.chain.IsSequenceStartJob(nextJob.Id) {
+			// Starting a sequence, so increment sequence try count.
+			r.chain.IncrementSequenceTries(nextJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+		}
+		r.runJobChan <- nextJob
+	}
+}
+
+// Finalize determines the final state of the chain and sends it to the Request
+// Manager.
+func (r *RunningChainReaper) Finalize(complete bool) {
+	if complete {
+		r.logger.Infof("chain is done, all jobs finished successfully")
+		r.chain.SetState(proto.STATE_COMPLETE)
+	} else {
+		r.logger.Warn("chain is done, some jobs failed")
+		r.chain.SetState(proto.STATE_FAIL)
+	}
+	r.chainRepo.Set(r.chain)
+	r.sendFinalState()
+	r.chainRepo.Remove(r.chain.RequestId())
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for chains that are being suspended (stopped to be resumed later).
+type SuspendedChainReaper struct {
+	reaper
+	runnerRepo runner.Repo
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if it's...
+// - completed: prepare subsequent jobs (copy jobData).
+// - failed:    prepare a sequence retry.
+// - stopped:   do nothing (job will be retried when chain is resumed).
+func (r *SuspendedChainReaper) Run() {
+	defer close(r.doneChan)
+
+	// Sleep a short time to prevent a race condition on the runner Repo. If a job
+	// was sent to runJobChan just before the chain was suspended, its runner might
+	// not have been created + added to the Repo yet. Wait so we can be sure all
+	// running jobs have runners in the Repo, so our checks to runnerRepo.Count
+	// will accurately reflect whether there are any running jobs left.
+	time.Sleep(runnerRepoWait)
```

Comment:
Would it be possible to not send jobs to runJobChan until their runners have been created and added to the Repo?

Reply:
It's *possible*, but I'd rather not. It would mean moving logic about running the jobs into the reaper, which should only be dealing with jobs after they're done running.

Not sure if the comment makes it clear, but sending the job to runJobChan is what causes the runner to be created (the loop in the traverser that reads from runJobChan creates the runners) - it just doesn't happen instantly, hence the wait.

Reply:
After reviewing the traverser this makes sense now. What are the implications if the sleep does not prevent the race condition, due to something like a GC pause, network latency, etc. (as Bill says, "One in a million is next Tuesday")?

Reply:
In that case, the reaper would return right away (per the `if runnerRepo.Count == 0` at line 309), without sending the final chain state or suspended job chain back to the RM. That's not great - the RM would think the request is still Running, even though the JR running it has shut down (basically, the same thing that happens if you shut down the current version of the JR while it's running a job chain).

We could increase the length of the sleep to make that less likely, but it's not a perfect solution.

Reply:
It's fine for now, but let's make a ticket to revisit and try to improve this later